bftengine/src/bftengine/CryptoManager.cpp

```diff
@@ -0,0 +1,245 @@
+    : cryptosys_(std::move(cs)) {}
+
+void CryptoManager::CryptoSystemWrapper::init() {
+  std::uint16_t f{ReplicaConfig::instance().getfVal()};
+  std::uint16_t c{ReplicaConfig::instance().getcVal()};
+  std::uint16_t numSigners{ReplicaConfig::instance().getnumReplicas()};
+  thresholdSigner_.reset(cryptosys_->createThresholdSigner());
+  thresholdVerifierForSlowPathCommit_.reset(cryptosys_->createThresholdVerifier(f * 2 + c + 1));
+  thresholdVerifierForCommit_.reset(cryptosys_->createThresholdVerifier(f * 3 + c + 1));
+  thresholdVerifierForOptimisticCommit_.reset(cryptosys_->createThresholdVerifier(numSigners));
+}
+
+CheckpointNum CryptoManager::getCheckpointOfCryptosystemForSeq(const SeqNum sn) const {
+  // find last chckp that is less than a chckp of a given sn
+  const uint64_t checkpointUpperBound = sn == 0 ? 0 : ((sn - 1) / checkpointWindowSize);
+  for (auto riter = checkpointToSystem().rbegin(); riter != checkpointToSystem().rend(); ++riter) {
+    auto& [checkpointCandidate, cryptosystem] = *riter;
+    if (checkpointCandidate <= checkpointUpperBound) {
+      LOG_DEBUG(logger(),
+                "Found cryptosystem for " << KVLOG(sn,
+                                                   checkpointUpperBound,
+                                                   checkpointCandidate,
+                                                   cryptosystem,
+                                                   checkpointToSystem().size(),
+                                                   checkpointWindowSize));
+      return checkpointCandidate;
+    }
+  }
+
+  // Replicas might encounter old messages, crashing is not
+  auto fallbackSystemCheckpoint = checkpointToSystem().rbegin()->first;
+  LOG_WARN(logger(),
+           "Cryptosystem not found, returning latest"
+               << KVLOG(sn, checkpointUpperBound, checkpointToSystem().size(), fallbackSystemCheckpoint));
+  return fallbackSystemCheckpoint;
+  // ConcordAssert(false && "should never reach here");
+}
+
+std::shared_ptr<CryptoManager::CryptoSystemWrapper> CryptoManager::get(const SeqNum& sn) const {
+  std::lock_guard<std::mutex> guard(mutex_);
+  return checkpointToSystem().at(getCheckpointOfCryptosystemForSeq(sn));
+}
+
+std::shared_ptr<CryptoManager::CryptoSystemWrapper> CryptoManager::create(const SeqNum& sn) {
+  std::lock_guard<std::mutex> guard(mutex_);
+  assertMapSizeValid();
+  // Cryptosystem for this sn will be activated upon reaching a second checkpoint from now
+  uint64_t chckp = (sn / checkpointWindowSize) + 2;
+  if (auto it = checkpointToSystem().find(chckp); it != checkpointToSystem().end()) {
+    LOG_WARN(logger(), "Cryptosystem already exists for checkpoint " << chckp);
+    return it->second;
+  }
+
+  // copy construct new Cryptosystem from a last one as we want it to include all the existing keys
+  std::unique_ptr<Cryptosystem> cs =
+      std::make_unique<Cryptosystem>(*checkpointToSystem().rbegin()->second->cryptosys_.get());
```

Comment:
Will we always create the cryptosystem in ascending order?
If not, wouldn't it make sense to take the cryptosystem of the largest sequence number that is still smaller than the given sequence number?